crates/solidity/outputs/cargo/crate/src/generated/bindings/generated/built_ins/0.8.7.sol

```diff
@@ -0,0 +1,101 @@
+// This file is generated automatically by infrastructure scripts. Please don't edit by hand.
+
+contract $BuiltIns$ {
+  function addmod(uint x, uint y, uint k) public returns (uint);
+  function assert(bool condition) public;
+  function blockhash(uint blockNumber) public returns (bytes32);
+  function ecrecover(bytes32 hash, uint8 v, bytes32 r, bytes32 s) public returns (address);
+  function gasleft() public returns (uint256);
+  function keccak256(bytes memory) public returns (bytes32);
+  function mulmod(uint x, uint y, uint k) public returns (uint);
+  function require(bool condition) public;
+  function require(bool condition, string memory message) public;
+  function revert() public;
+  function revert(string memory reason) public;
+  function ripemd160(bytes memory) public returns (bytes20);
+  function selfdestruct(address payable recipient) public;
+  function sha256(bytes memory) public returns (bytes32);
+  struct $abiType {
+    function(bytes memory, $args) returns ($args) decode;
+    function($args) returns (bytes memory) encode;
+    function($args) returns (bytes memory) encodePacked;
+    function(bytes4 selector, $args) returns (bytes memory) encodeWithSelector;
+    function(string memory, $args) returns (bytes memory) encodeWithSignature;
+  }
+  struct $address {
+    uint256 balance;
+    bytes code;
+    bytes32 codehash;
+    function(bytes memory) returns (bool, bytes memory) call;
+    function(bytes memory) returns (bool, bytes memory) delegatecall;
+    function(uint256) returns (bool) send;
+    function(bytes memory) returns (bool, bytes memory) staticcall;
+    function(uint256) transfer;
+  }
+  struct $array {
+    uint length;
+    function(uint) returns ($element) $index;
+    function() returns ($element) push;
+    function($element) push;
+    function() pop;
+  }
+  struct $arrayFixed {
+    uint length;
+    function(uint) returns ($element) $index;
+  }
+  struct $blockType {
+    uint basefee;
+    uint chainid;
+    address payable coinbase;
+    uint difficulty;
+    uint gaslimit;
+    uint number;
+    uint timestamp;
+  }
+  struct $bytes {
+    function($args) returns (bytes memory) concat;
+  }
+  struct $callOptions {
+    uint gas;
+    uint salt;
+    uint value;
+  }
+  struct $functionExternal {
+    $address $address;
+    $selector selector;
+  }
+  struct $msgType {
+    bytes data;
+    address sender;
+    bytes4 sig;
+    uint value;
+  }
+  struct $string {
+    function($args) returns (string memory) concat;
+  }
+  struct $txType {
+    uint gasprice;
+    address origin;
+  }
+  struct $typeContractType {
+    string name;
+    bytes creationCode;
+    bytes runtimeCode;
+    bytes4 interfaceId;
+  }
+  struct $typeInterfaceType {
+    string name;
+    bytes4 interfaceId;
+  }
+  struct $typeIntType {
+    int min;
+    int max;
+  }
+  $function $placeholder;
+  $abiType abi;
+  $blockType block;
+  $msgType msg;
+  $SuperType super;
+  $ThisType this;
```

Comment:
for `this`/`super`, I wonder why they are different from normal instance variables?
`this` would be resolved as a reference to the enclosing contract, and `super` can resolve to ALL parent contracts (similar to virtual method calls).
This way, their property access/types/etc... don't need special handling.

Reply:
I'm not sure about this proposed change. Both `this` and `super` refer to the instances of the contracts, not the types. I agree it may be useful but it's not correct (and can potentially lead to incorrect bindings, eg. `this.Foo` would resolve to a `struct Foo` defined in the contract). For reference, Remix doesn't resolve either of them.

Reply:
This is now implemented as you suggested. The recent change converting both `super` and `this` to keywords meant that they no longer can be used as variables, so the above code would no longer parse. There is one minor caveat to this change, related to performance. Because to resolve to the contract or parents we need to push their symbols when resolving, this provides a branching path in the stack graph to account for this possibility. It should only affect resolution stacks involving `this` or `super` but I've become more aware of adding new branching since we've started having performance issues. Hopefully this change does not have too much of an impact.

Reply:
> Both `this` and `super` refer to the instances of the contracts

Sorry for the confusion. This is exactly what I meant. We have these "global variables" here, to let the "normal" binding rules bind such identifiers to an instance of the contracts. But this results in us adding/exposing a global variable here as a built-in, even if it is not (it is a language keyword, not an actual global variable exported from a virtual `built_ins.sol` file).

My suggestion was, if we are using a separate rules file for built ins, can we just do this binding using a rule for these two keywords specifically, to bind them to the instance (not the type) directly, rather than having this simulated through `$SuperType super` and `$ThisType this` here? The direct benefit is that we don't expose this hack to users.

Reply:
> Both `this` and `super` refer to the instances of the contracts

Sorry for the confusion. This is exactly what I meant. We have these "global variables" here, to let the "normal" binding rules bind such identifiers to an instance of the contracts. But this results in us adding/exposing a global variable here as a built-in, even if it is not (it is a language keyword, not an actual global variable exported from a virtual `built_ins.sol` file).

My suggestion was, if we are using a separate rules file for built ins, can we just do this binding using a rule for these two keywords specifically, to bind them directly, rather than having this simulated through `$SuperType super` and `$ThisType this` here? The direct benefit is that we don't expose this hack to users.

Reply:
The rules for binding both are now integrated in the `ContractDefinition` and `InheritanceSpecifier` which is where we have direct access to the contract and its parents. They can be separated into another file for readability when we split the rules into multiple files though.